src/overlays/actors/ovl_Obj_Y2shutter/z_obj_y2shutter.c

```diff
@@ -28,24 +41,157 @@ const ActorInit Obj_Y2shutter_InitVars = {
+        }
+        if (Flags_GetSwitch(play, this->dyna.actor.params & 0x7F)) {
+            Flags_UnsetSwitch(play, this->dyna.actor.params & 0x7F);
+        } else {
+            Flags_SetSwitch(play, this->dyna.actor.params & 0x7F);
+        }
+    }
+    if (this->unk_15F == 0) {
+        if (Flags_GetSwitch(play, this->dyna.actor.params & 0x7F)) {
+            s16 cutscene = this->dyna.actor.cutscene;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Y2shutter/func_80B9AA20.s")
+            if (this->unk_15C == 0) {
+                if ((cutscene >= 0) && (!ActorCutscene_GetCanPlayNext(cutscene))) {
+                    ActorCutscene_SetIntentToPlay(cutscene);
+                } else if (cutscene >= 0) {
+                    ActorCutscene_StartAndSetUnkLinkFields(cutscene, &this->dyna.actor);
+                    this->unk_15C = -1;
+                } else {
+                    func_80B9AA20(this, info, shutterType);
+                }
+            } else if (this->unk_15C < 0) {
+                if (func_800F22C4(cutscene, &this->dyna.actor)) {
+                    func_80B9AA20(this, info, shutterType);
+                }
+            } else {
+                sp2C = this->dyna.actor.home.pos.y + info->unk_8;
+                sp28 = info->unk_C;
+                sp24 = info->unk_10;
+                if (this->unk_15C < 2) {
+                    Flags_UnsetSwitch(play, this->dyna.actor.params & 0x7F);
+                } else {
+                    this->unk_15C--;
+                }
+            }
+        } else if (this->unk_15C != 0) {
+            this->unk_15C = 0;
+            this->unk_15F = info->unk_1C;
+            if (shutterType == 0) {
+                Actor_PlaySfxAtPos(&this->dyna.actor, NA_SE_EV_METALDOOR_CLOSE);
+            }
+        } else {
+            sp2C = this->dyna.actor.home.pos.y;
+            sp28 = info->unk_14;
+            sp24 = info->unk_18;
+        }
+    }
+    Math_StepToF(&this->dyna.actor.velocity.y, sp28, sp24);
+    this->dyna.actor.world.pos.y += this->dyna.actor.velocity.y;
+    if (((this->dyna.actor.world.pos.y - sp2C) * sp28) >= 0.0f) {
+        this->dyna.actor.world.pos.y = sp2C;
+        this->dyna.actor.velocity.y = 0.0f;
+        if (!this->unk_15E) {
+            this->unk_15E = true;
+            this->unk_15F = info->unk_1D;
+            if (shutterType != 0) {
+                Actor_PlaySfxAtPos(&this->dyna.actor, NA_SE_EV_METALDOOR_STOP);
+            }
+        }
+    } else {
+        this->unk_15E = false;
+        if (shutterType != 0) {
+            func_800B9010(&this->dyna.actor, 0x2036U);
+        }
+    }
+    if (DECR(this->unk_15F) != 0) {
+        this->dyna.actor.shape.yOffset = 2.0f * (this->unk_15F & 1) * this->unk_15F;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Y2shutter/ObjY2shutter_Update.s")
+void ObjY2shutter_Draw(Actor* thisx, PlayState* play) {
+    ShutterInfo* info;
+    ObjY2shutter* this = THIS;
```

Comment:
Does the stack in this and the other functions have to go in this order? I.e. is it possible to have THIS first (possibly with another pad above)?

Reply:
moving THIS to the top matches in every function except Init, it seems